analyses/cell-type-ewings/template_notebooks/clustering-workflow/01-clustering-metrics.Rmd

```diff
@@ -0,0 +1,257 @@
+
+
+```{r packages}
+suppressPackageStartupMessages({
+  # load required packages
+  library(SingleCellExperiment)
+  library(ggplot2)
+})
+
+# Set default ggplot theme
+theme_set(
+  theme_classic()
+)
+
+# set seed
+set.seed(2024)
+```
+
+
+```{r base paths}
+# The path to this module
+module_base <- rprojroot::find_root(rprojroot::is_renv_project)
+
+# source in clustering functions
+clustering_functions <- file.path(module_base, "scripts", "utils", "clustering-functions.R")
+source(clustering_functions)
+```
+
+
+```{r}
+# read in input sce files
+sce <- readr::read_rds(params$sce_file)
+```
+
+
+## Calculate clusters
+
+
+```{r}
+# louvain clustering across all parameters 
+louvain_results <- rOpenScPCA::sweep_clusters(
+  sce,
+  algorithm = "louvain",
+  nn = seq(5, 40, 5),
+  resolution = c(0.5, 1, 1.5)
+) |> 
+  dplyr::bind_rows() |> 
+  # add in objective function so we can combine with leiden results
+  dplyr::mutate(objective_function = NA_character_)
+
+# leiden cpm clustering
+leiden_cpm_results <- rOpenScPCA::sweep_clusters(
+  sce,
+  algorithm = "leiden",
+  nn = seq(5, 40, 5),
+  resolution = c(0.001, .005, .01),
+  objective_function = "CPM"
+) |> 
+  dplyr::bind_rows()
+
+# leiden modularity clustering
+leiden_mod_results <- rOpenScPCA::sweep_clusters(
+  sce,
+  algorithm = "leiden",
+  nn = seq(5, 40, 5),
+  resolution = c(0.5, 1, 1.5),
+  objective_function = "modularity"
+) |> 
+  dplyr::bind_rows()
+```
+
+
+```{r}
+# join all clustering results 
+all_cluster_results <- dplyr::bind_rows(list(louvain_results, leiden_cpm_results, leiden_mod_results)) |> 
+  # add column for method + objective function 
+  # use this for splitting plots throughout report
+  dplyr::mutate(cluster_method = glue::glue(
+    "{algorithm}-{objective_function}"
+  ) |> 
+    stringr::str_remove("-NA")
+  )
+```
+
+
+## Visualize clusters with UMAP
+
+```{r}
+# get umap embeddings and combine into a data frame with cluster assignments
+umap_df <- sce |>
+  scuttle::makePerCellDF(use.dimred = "UMAP") |>
+  # replace UMAP.1 with UMAP1 and get rid of excess columns
+  dplyr::select("cell_id" = barcodes, UMAP1 = UMAP.1, UMAP2 = UMAP.2) |>
+  dplyr::left_join(all_cluster_results, by = "cell_id")
+
+split_umap_df <- umap_df |> 
+  # add a column that shows the combination of nn/res so we get stats for each unique combo of params 
+  dplyr::mutate(cluster_params = glue::glue("{nn}nn-{resolution}res")) |> 
+  split(umap_df$cluster_method)
+```
+
+
+Below we visualize the cluster assignments using each parameter on a UMAP. 
+This can be helpful to see if any parameters lead to any obvious over or under clustering. 
+
+```{r, fig.height=15, fig.width=10}
+# look at clustering results for each library across all params
+split_umap_df |> 
+  purrr::imap(\(df, method){
+    
+    ggplot(df, aes(x = UMAP1, y = UMAP2, color = cluster)) +
+      geom_point(alpha = 0.5, size = 0.1) +
+      facet_grid(rows = vars(nn),
+                 cols = vars(resolution),
```

Comment:
I think I might actually flip these so the overall look is more horizontal? You may also need to change the r chunk fig width/height.

Reply:
Since there are more `nn` values used than resolution I think having a wider plot would be hard and make the UMAPs _really_ small. So I'm not super inclined to change this.